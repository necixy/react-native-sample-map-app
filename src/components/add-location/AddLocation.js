import React, { Component } from "react";
import {
  View,
  Text,
  TextInput,
  FlatList,
  TouchableOpacity
} from "react-native";
import KEYS from "../../config/keys";
import { graphql } from "react-apollo";
import locationsQuery from "../../graphql/queries/locations";
import addLocation from "../../graphql/mutations/addLocation";
import s from "./AddLocationStyles";

class AddLocation extends Component {
  static navigationOptions = {
    title: "Add Location"
  };

  constructor(props) {
    super(props);
    this.state = {
      searchText: "",
      locations: null,
      error_message: null
    };
  }

  async geocodeAddress() {
    const { searchText } = this.state;
    if (!searchText) {
      this.setState({
        locations: null,
        error_message: "Please type in the address in order to search."
      });
      return;
    }
    const { GOOGLE_GEOCODING_KEY } = KEYS;
    const response = await fetch(
      `https://maps.googleapis.com/maps/api/geocode/json?&address=${searchText}&key=${GOOGLE_GEOCODING_KEY}`
    );
    const data = await response.json();
    this.setGeoCodingResults(data);
  }

  setGeoCodingResults(data) {
    if (data.status == "OK") {
      this.setState({ locations: data.results, error_message: null });
    } else if (data.status == "ZERO_RESULTS") {
      this.setState({ locations: null, error_message: "No results found" });
    } else {
      this.setState({ locations: null, error_message: data.error_message });
    }
  }
  async addLocation(location) {
    await this.props.mutate({
      variables: {
        name: this.state.searchText,
        address: location.formatted_address,
        lat: location.geometry.location.lat,
        lng: location.geometry.location.lng
      },
      refetchQueries: [{ query: locationsQuery }]
    });
    this.props.navigation.pop();
  }

  renderErrorMsg() {
    const { error_message } = this.state;
    return error_message ? (
      <Text style={s.errorMsg}>{error_message}</Text>
    ) : null;
  }

  renderListItem({ item, index }) {
    return (
      <TouchableOpacity
        key={index}
        style={s.listItem}
        onPress={this.addLocation.bind(this, item)}
      >
        <Text testId="textAddress">Address: {item.formatted_address}</Text>
        <Text testId="textLat">Latitude: {item.geometry.location.lat}</Text>
        <Text testId="textLng">Longitude: {item.geometry.location.lng}</Text>
      </TouchableOpacity>
    );
  }

  renderList() {
    const _keyExtractor = (item, index) => `location-${index}`;
    const { locations } = this.state;
    return locations == null ? null : (
      <View style={s.listContainer}>
        <View style={s.listHeader}>
          <Text style={s.listHeaderLbl}>
            Please tap on the desired address below to add to the map.
          </Text>
        </View>
        <FlatList
          textId="locationsList"
          style={s.list}
          data={locations}
          keyExtractor={_keyExtractor}
          renderItem={this.renderListItem.bind(this)}
        />
      </View>
    );
  }

  render() {
    return (
      <View style={s.container}>
        <TextInput
          testId="searchInput"
          placeholder="Type in the address"
          onChangeText={async text => {
            await this.setState({ searchText: text });
          }}
          value={this.state.searchText}
          returnKeyType="search"
          autoFocus={true}
          style={s.addressInput}
          onSubmitEditing={this.geocodeAddress.bind(this)}
        />
        {this.renderErrorMsg()}
        {this.renderList()}
      </View>
    );
  }
}

export default graphql(addLocation)(AddLocation);
