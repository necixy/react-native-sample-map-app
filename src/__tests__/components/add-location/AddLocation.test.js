import "react-native";
import React from "react";
import renderer from "react-test-renderer";
import { MockedProvider } from "react-apollo/test-utils";
import AddLocation from "../../../components/add-location/AddLocation";

const mockedGeocodingResult = {
  results: [
    {
      address_components: [
        {
          long_name: "Berlin",
          short_name: "Berlin",
          types: ["locality", "political"]
        },
        {
          long_name: "Berlin",
          short_name: "Berlin",
          types: ["administrative_area_level_1", "political"]
        },
        {
          long_name: "Germany",
          short_name: "DE",
          types: ["country", "political"]
        }
      ],
      formatted_address: "Berlin, Germany",
      geometry: {
        bounds: {
          northeast: {
            lat: 52.6754542,
            lng: 13.7611175
          },
          southwest: {
            lat: 52.338234,
            lng: 13.088346
          }
        },
        location: {
          lat: 52.52000659999999,
          lng: 13.404954
        },
        location_type: "APPROXIMATE",
        viewport: {
          northeast: {
            lat: 52.6754542,
            lng: 13.7611175
          },
          southwest: {
            lat: 52.338234,
            lng: 13.088346
          }
        }
      },
      place_id: "ChIJAVkDPzdOqEcRcDteW0YgIQQ",
      types: ["locality", "political"]
    }
  ],
  status: "OK"
};

describe("AddLocation Component Testing", () => {
  it("AddLocation Renders correctly", async () => {
    const component = renderer.create(
      <MockedProvider addTypename={false}>
        <AddLocation />
      </MockedProvider>
    );

    let root = component.root;
    let addLocationInstance =
      root.children[0].children[0].children[0].children[0].instance;

    expect(component.toJSON()).toMatchSnapshot();

    addLocationInstance.geocodeAddress = jest.fn(() => {
      addLocationInstance.setGeoCodingResults(mockedGeocodingResult);
    });

    let searchInput = root.find(
      el =>
        el.props && el.props.testId == "searchInput" && el.type == "TextInput"
    );
    expect(searchInput).toBeDefined();

    const searchString = "Berlin";
    await searchInput.props.onChangeText(searchString);
    await searchInput.props.onSubmitEditing();

    expect(component.toJSON()).toMatchSnapshot();

    const { searchText, locations, error_message } = addLocationInstance.state;
    expect(searchInput.props.value).toBe(searchString);
    expect(searchText).toBe(searchString);
    expect(locations).toBeDefined();
    expect(error_message).toBeNull();

    const textAddress = root.find(
      el => el.props && el.props.testId == "textAddress" && el.type == "Text"
    );
    const textLat = root.find(
      el => el.props && el.props.testId == "textLat" && el.type == "Text"
    );
    const textLng = root.find(
      el => el.props && el.props.testId == "textLng" && el.type == "Text"
    );
    const location = mockedGeocodingResult.results[0];
    expect(textAddress.children[1]).toBe(location.formatted_address);
    expect(textLat.children[1]).toBe(location.geometry.location.lat.toString());
    expect(textLng.children[1]).toBe(location.geometry.location.lng.toString());
  });
});
