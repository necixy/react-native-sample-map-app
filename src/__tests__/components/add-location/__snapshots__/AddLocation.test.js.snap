// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddLocation Component Testing AddLocation Renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <TextInput
    allowFontScaling={true}
    autoFocus={true}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Type in the address"
    returnKeyType="search"
    style={
      Object {
        "backgroundColor": "#f5f5f5",
        "padding": 20,
      }
    }
    testId="searchInput"
    underlineColorAndroid="transparent"
    value=""
  />
</View>
`;

exports[`AddLocation Component Testing AddLocation Renders correctly 2`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <TextInput
    allowFontScaling={true}
    autoFocus={true}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Type in the address"
    returnKeyType="search"
    style={
      Object {
        "backgroundColor": "#f5f5f5",
        "padding": 20,
      }
    }
    testId="searchInput"
    underlineColorAndroid="transparent"
    value="Berlin"
  />
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "borderBottomColor": "#ccc",
          "borderBottomWidth": 1,
          "padding": 20,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#777",
          }
        }
      >
        Please tap on the desired address below to add to the map.
      </Text>
    </View>
    <RCTScrollView
      data={
        Array [
          Object {
            "address_components": Array [
              Object {
                "long_name": "Berlin",
                "short_name": "Berlin",
                "types": Array [
                  "locality",
                  "political",
                ],
              },
              Object {
                "long_name": "Berlin",
                "short_name": "Berlin",
                "types": Array [
                  "administrative_area_level_1",
                  "political",
                ],
              },
              Object {
                "long_name": "Germany",
                "short_name": "DE",
                "types": Array [
                  "country",
                  "political",
                ],
              },
            ],
            "formatted_address": "Berlin, Germany",
            "geometry": Object {
              "bounds": Object {
                "northeast": Object {
                  "lat": 52.6754542,
                  "lng": 13.7611175,
                },
                "southwest": Object {
                  "lat": 52.338234,
                  "lng": 13.088346,
                },
              },
              "location": Object {
                "lat": 52.52000659999999,
                "lng": 13.404954,
              },
              "location_type": "APPROXIMATE",
              "viewport": Object {
                "northeast": Object {
                  "lat": 52.6754542,
                  "lng": 13.7611175,
                },
                "southwest": Object {
                  "lat": 52.338234,
                  "lng": 13.088346,
                },
              },
            },
            "place_id": "ChIJAVkDPzdOqEcRcDteW0YgIQQ",
            "types": Array [
              "locality",
              "political",
            ],
          },
        ]
      }
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      numColumns={1}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      style={
        Object {
          "flex": 1,
        }
      }
      textId="locationsList"
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "borderBottomColor": "#ccc",
                "borderBottomWidth": 1,
                "marginTop": 2,
                "minHeight": 50,
                "opacity": 1,
                "padding": 20,
              }
            }
          >
            <Text
              testId="textAddress"
            >
              Address: 
              Berlin, Germany
            </Text>
            <Text
              testId="textLat"
            >
              Latitude: 
              52.52000659999999
            </Text>
            <Text
              testId="textLng"
            >
              Longitude: 
              13.404954
            </Text>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
